### PR TITLE
Added shotcode and tokens for Automatic induction scheduling

### PR DESCRIPTION
**Pull request by nishant22029**
- Created a new token for the email message template to include the induction slot date and time, as well as the online induction meeting link, for sending slot booking details via email.
- Added a shortcode to capture source_contact_id, slot_date, slot_time, and induction_type from URL parameters, allowing dynamic handling of induction slot bookings.
Updated the contact induction activity's date, time, and status to "Scheduled" if not already scheduled or completed.
Selected the appropriate email template based on induction type to send acknowledgment emails after slot booking.

## Summary by CodeRabbit

- **New Features**
	- Introduced new tokens for induction processes, including `inductionDateTime` and `inductionOnlineMeetlink`.
	- Added a shortcode for managing induction slot details, allowing for scheduling and email notifications based on induction activities.

- **Bug Fixes**
	- Enhanced error handling for invalid date/time formats and missing email templates during induction slot management.

### CODE REVIEW

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 1]
```diff
@@ -65,9 +65,12 @@ function goonjcustom_civicrm_container(ContainerBuilder $container) {
  */
 function goonjcustom_register_tokens(TokenRegisterEvent $e) {
   $e->entity('contact')
-    ->register('inductionDetails', ts('Induction details'));
+    ->register('inductionDetails', ts('Induction details'))
+    ->register('inductionDateTime', ts('Induction Scheduled Date & Time'))
+    ->register('inductionOnlineMeetlink', ts('Induction Online Meet link'));
 }
 
+
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:73]: Remove extra space

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 2]
```diff
@@ -127,6 +130,35 @@ function goonjcustom_evaluate_tokens(TokenValueEvent $e) {
...
 
       $row->tokens('contact', 'inductionDetails', $inductionDetailsMarkup);
+			// Fetch induction activity details
+			$inductionActivity = \Civi\Api4\Activity::get(FALSE)
+					->addSelect('activity_date_time', 'Induction_Fields.Goonj_Office')
+					->addWhere('activity_type_id:name', '=', 'Induction')
+					->addWhere('source_contact_id', '=', $contactId)
+					->setLimit(1)
+					->execute()
+					->first();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:140]: Use single everywhere on this PR wherever needed

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 3]
```diff
@@ -127,6 +130,35 @@ function goonjcustom_evaluate_tokens(TokenValueEvent $e) {
...
+
+			// Fetch office online meet link details if induction office is specified
+			$inductionOnlineMeetlink = '';
+			if ($inductionGoonjOffice) {
+					$officeDetails = \Civi\Api4\Contact::get(FALSE)
+							->addSelect('Goonj_Office_Details.Induction_Meeting_Access_Link')
+							->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+							->addWhere('id', '=', $inductionGoonjOffice)
+							->execute()
+							->first();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:153]: Here

File: wp-content/themes/goonj-crm/engine/shortcodes.php [hunk 4]
```diff
@@ -130,3 +132,86 @@ function goonj_collection_camp_past_data() {
...
+
+    try {
+        // Fetch the induction activity for the specified source contact
+        $inductionActivity = \Civi\Api4\Activity::get(FALSE)
+            ->addSelect('id', 'activity_date_time', 'status_id:name', 'Induction_Fields.Goonj_Office', 'Induction_Fields.Assign')
+            ->addWhere('source_contact_id', '=', $source_contact_id)
+            ->addWhere('activity_type_id:name', '=', 'Induction')
+            ->execute()
+            ->single();
+		\Civi::log()->info('inductionActivity', ['inductionActivity'=>$inductionActivity]);
```
Comment (tarunnjoshi) [inline wp-content/themes/goonj-crm/engine/shortcodes.php:151]: Can we remove the logs or we need this?

File: wp-content/themes/goonj-crm/engine/shortcodes.php [hunk 5]
```diff
@@ -130,3 +132,86 @@ function goonj_collection_camp_past_data() {
...
+
+
+        // Update activity date and time and set the status to "Scheduled"
+        \Civi\Api4\Activity::update(FALSE)
+            ->addValue('activity_date_time', $newActivityDateTime->format('Y-m-d H:i:s'))
+            ->addValue('status_id:name', 'Scheduled')
+            ->addValue('Induction_Fields.Mode:name', $inductionType)
+            ->addWhere('id', '=', $inductionActivity['id'])
+            ->execute();
+        \Civi::log()->info('Activity updated successfully', [
```
Comment (tarunnjoshi) [inline wp-content/themes/goonj-crm/engine/shortcodes.php:174]: Needed?

File: wp-content/themes/goonj-crm/engine/shortcodes.php [hunk 6]
```diff
@@ -130,3 +132,86 @@ function goonj_collection_camp_past_data() {
...
+        // Select the email template based on the induction type
+        $templateTitle = ($inductionType == 'Processing_Unit') 
+            ? 'Acknowledgment_for_Induction_Slot_Booked' 
+            : 'Acknowledgment_for_Online_Induction_Slot_Booked';
+
+        // Fetch the email template
+        $template = \Civi\Api4\MessageTemplate::get(FALSE)
+            ->addWhere('msg_title', 'LIKE', $templateTitle . '%')
+            ->execute()
+            ->single();
```
Comment (tarunnjoshi) [inline wp-content/themes/goonj-crm/engine/shortcodes.php:188]:
```suggestion
            ->execute()->single();
```

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php (hunk 2 above)
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:140]:
```suggestion
					->execute()->first();
```

File: wp-content/themes/goonj-crm/engine/shortcodes.php [hunk 7]
```diff
@@ -130,3 +132,86 @@ function goonj_collection_camp_past_data() {
...
+            $emailParams = [
+                'contact_id' => $source_contact_id,
+                'template_id' => $template['id'],
+                'cc' => $assigneeContact['email.email']
+            ];
+
+            // Send the email
+            $emailResult = civicrm_api3('Email', 'send', $emailParams);
+        } else {
+            \Civi::log()->error('No email template found', ['template_title' => $templateTitle]);
```
Comment (tarunnjoshi) [inline wp-content/themes/goonj-crm/engine/shortcodes.php:209]: Any usecase?

**Review by tarunnjoshi — CHANGES_REQUESTED** [review]

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php (hunk 1 above)
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/goonjcustom.php:73]: sure

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php (hunk 2 above)
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/goonjcustom.php:140]: sure

File: wp-content/themes/goonj-crm/engine/shortcodes.php (hunk 4 above)
Comment (nishant22029) [reply wp-content/themes/goonj-crm/engine/shortcodes.php:151]: sure

File: wp-content/themes/goonj-crm/engine/shortcodes.php (hunk 5 above)
Comment (nishant22029) [reply wp-content/themes/goonj-crm/engine/shortcodes.php:174]: Yes, for monitoring logs or any bug occur we can check this log

File: wp-content/themes/goonj-crm/engine/shortcodes.php (hunk 7 above)
Comment (nishant22029) [reply wp-content/themes/goonj-crm/engine/shortcodes.php:209]: removed

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 8]
```diff
@@ -127,6 +129,34 @@ function goonjcustom_evaluate_tokens(TokenValueEvent $e) {
       }
 
       $row->tokens('contact', 'inductionDetails', $inductionDetailsMarkup);
+			// Fetch induction activity details
+			$inductionActivity = \Civi\Api4\Activity::get(FALSE)
+					->addSelect('activity_date_time', 'Induction_Fields.Goonj_Office')
+					->addWhere('activity_type_id:name', '=', 'Induction')
+					->addWhere('source_contact_id', '=', $contactId)
+					->execute()
+					->single();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:138]:
```suggestion
					->execute()->single();
```

File: wp-content/civi-extensions/goonjcustom/goonjcustom.php [hunk 9]
```diff
@@ -127,6 +129,34 @@ function goonjcustom_evaluate_tokens(TokenValueEvent $e) {
...
+
+			// Fetch office online meet link details if induction office is specified
+			$inductionOnlineMeetlink = '';
+			if ($inductionGoonjOffice) {
+					$officeDetails = \Civi\Api4\Contact::get(FALSE)
+							->addSelect('Goonj_Office_Details.Induction_Meeting_Access_Link')
+							->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+							->addWhere('id', '=', $inductionGoonjOffice)
+							->execute()
+              ->single();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/goonjcustom.php:151]:
```suggestion
							->execute()->single();
```
Can we do this all the places

File: wp-content/themes/goonj-crm/engine/shortcodes.php [hunk 10]
```diff
@@ -130,3 +132,81 @@ function goonj_collection_camp_past_data() {
...
+    $inductionType = $_GET['induction_type'] ?? '';
+
+    try {
+        // Fetch the induction activity for the specified source contact
+        $inductionActivity = \Civi\Api4\Activity::get(FALSE)
+            ->addSelect('id', 'activity_date_time', 'status_id:name', 'Induction_Fields.Goonj_Office', 'Induction_Fields.Assign')
+            ->addWhere('source_contact_id', '=', $source_contact_id)
+            ->addWhere('activity_type_id:name', '=', 'Induction')
+            ->execute()
+            ->single();
```
Comment (tarunnjoshi) [inline wp-content/themes/goonj-crm/engine/shortcodes.php:150]: Herer

**Review by tarunnjoshi — APPROVED** [review]

**Pull request merged**